### ISSUE TITLE
[Experiment] - We should prefer switching to open tabs vs opening a duplicate tab

### ISSUE THREAD

**Issue by Verdi** [mozilla-mobile/firefox-ios#15236]
Hypothesis: It is better to switch tabs instead of opening a duplicate tab (Avoid duplication and increase app performance)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-6836)

**Comment by OrlaM** [comment]
Tabs and bookmarks/shortcuts are very different concepts (both in code and to a user). A tab will have some amount of state associated with it based on how the user previously interacted with it that may not be desirable to a user who is trying to open a new page. Desktop doesn't treat bookmarks or shortcuts this way, does Android?
This is especially potentially confusing considering the homepage on iOS is a tab. 
Consider the following work flow: 
- Browse a page like amazon.com
- Open the tab tray
- Select new tab
- Select a shortcut for amazon.com
- Be brought to the previous amazon.com tab which is in the same state it was left in

Even if we do exact matching on the URL and only go to an existing tab for a 100% match it would still be problematic as a lot of modern pages are just javascript and the url doesn't change when exploring parts of a site.

I'm not saying we shouldn't do this just because other platforms don't but this is something that should at least be run as an experiment if there is a hypothisis that the existing behaviour is causing friction for the user. That way we can validate without risking upsetting the entire user base, they can get really tetchy when we mess with this type of thing.

Also this isn't a bug, it's a feature request.

**Comment by data-sync-user** [comment]
➤ Roux Buciu commented:

Michael Verdi is gonna do the sync dance w/ Daniela Arcese

**Comment by chemicalcain** [comment]
Agree with OrlaM that this deviates from standard expected behavior of bookmarks. I may have multiple tabs open to a "New entry" page on my blog which is the same link, but with different text in progress in each tab, for example. Clicking a bookmark for that page should open as a new tab. Additionally, as OrlaM says, many sites are 100% Javascript-- how would this change handle page anchors and existing cached user-entered data?

A setting for this behavior might be reasonable if it proceeds past the test stage.

**Comment by data-sync-user** [comment]
➤ Bridget Kaluzny commented:

Jeremy Evans    to check on the effort here please,

**Comment by data-sync-user** [comment]
➤ Matthew Reagan commented:

Taking a quick look to see if I could pick up this task; based on comments below it looks like there may still be open questions / UX concerns on this ticket that need to be clarified further?

Can anyone confirm if these have been resolved, and if so what the action items are? (cc Orla Mitchell Michael Verdi) Thank you!

**Comment by data-sync-user** [comment]
➤ Lauren Reed commented:

Hey y’all !
I went digging here and found that Android has already[ implemented a similar feature|https://bugzilla.mozilla.org/show_bug.cgi?id=1840721] here. 

Android implemented: 

  * Pinned user shortcuts (user created pins) open in an existing tab if it has the same URL  
    * Sponsored/default top sites always open a new tab and do not attempt to URL match for an open tab 

Orla Mitchell Michael Verdi does matching the behavior from Android resolve the concerns listed earlier in this ticket, or do we want to explore a setting or an experiment to get this feature moving? 

I believe Tom White implemented this on android, and might be a good resource should we have additional questions!

**Comment by data-sync-user** [comment]
➤ Michael Verdi commented:

Having the items on the homepage open existing tabs instead of new tabs is a fairly common request from users as well as being an issue that the UX team identified. Android implemented this exact same request in Firefox 118. I think Lauren’s suggestion of talking to Tom about implementation details is helpful. I still think this is something we want to do and see no reason why this should work differently on iOS.

**Comment by data-sync-user** [comment]
➤ Bridget Kaluzny commented:

Lauren meeting with Andy today 11/29, and targeting 122 release.

**Comment by data-sync-user** [comment]
➤ Norberto Andres Furlan commented:

Michael Verdi can you confirm? Thanks!

**Comment by data-sync-user** [comment]
➤ Michael Verdi commented:

Yes this is correct.

**Comment by data-sync-user** [comment]
➤ Michael Verdi commented:

I think most recently accessed would make most sense.

**Comment by data-sync-user** [comment]
➤ Nishant Bhasin commented:

Hi, it would be much easier to just go to the tab with the same url regardless of most recent. 

As most recent could be the tab at the bottom or the first ever tab we opened many eons ago.

Reason for this is I am trying to avoid any special sort operation to simplify the logic. And in most cases a user will go to the most recent tab if we just loop from the last opened tab to the first opened tab and if tab which matches the same URL we go there

**Comment by data-sync-user** [comment]
➤ Michael Verdi commented:

Nishant Bhasin I guess then my question is technically, what is the most recent tab? Can you explain how this would be: “As most recent could be the tab at the bottom or the first ever tab we opened many eons ago”

**Comment by data-sync-user** [comment]
➤ Nishant Bhasin commented:

Depends on the definition of most recent tab. 
Lets assume you opened the following tabs in the order mentioned below

1 → 2 → 3 → 4 → 5 

1st being the oldest tab you opened and 5 being the most recent tab you opened. Now a user may interact with 1st tab more often. Then would the 1st tab become most recent or the 5th? 

In most cases users just opens a new tab on mobile or interacts with their last tab which is 5th tab in the order mentioned above. 

Keeping that in mind I am suggesting we skip the whole most recent and simply loop from last tab that the user Created and check if the url matches. If it doesn’t then we add a new tab 6th or if the url exists in 5th or 4th we switch to that. 

Special case: Where both 5th and 4th are tab with same url that we want to go then since we are going from Created scenario and checking 5th then checking 4th means it will take the user to the 5th tab.

tl;dr → We loop backwards from your list of tabs in tab tray and stop as soon as we find the tab with the url and if not then we add a new tab

Michael Verdi

**Comment by data-sync-user** [comment]
➤ Norberto Andres Furlan commented:

Alignment: Before doing the change for all the sections in the homepage, we are going to run an experiment with the shortcuts.

**Comment by data-sync-user** [comment]
➤ Razvan Litianu commented:

hey Norberto Andres Furlan on closer inspection it seems that we don’t open a new tab for the Shortcuts, Recently Visited and Recently Saved Section. On iOS we have the Homescreen tab that acts as a real tab. Tapping on any link from the sections above will result in loading the url inside the Homescreen tab. You can notice that you can actually go back home and back on the page using back/forward buttons. And you can do this multiple times, opening a shortcut and then tapping home then opening a recently saved link and you will have to option to back an forth in between these pages. The only way we open a new tab for these is to tap and hold and selecting from the menu Open in new tab.

So my question would be if we’re not opening a new tab for these cases, should we change the cases?

**Comment by data-sync-user** [comment]
➤ Razvan Litianu commented:

Hello, coming back with an update about tabs. I did an investigation on Android and it seems we have completely different ways of navigating from what is considered the Homescreen. As previously mentioned iOS is considering home as tab and all the links are opened inside the same tab. Android opens every link in a new tab so this change seems to make more sense on their side. The team told me there is also a plan to refactor home and have the behaviour be close to what Android does and we think it is best to postpone this task until that refactor is made.

I’m attaching videos on how the apps open links from home, notice the tab count.

Tagging Roux Buciu and Laurie Marceau for visibility.

AndroidiOS!XRecorder05012024114508.mp4width=720,height=1600!!Simulator Screen Recording - iPhone 15 - 20240105 at 12.56.02.mp4width=1178,height=2556!

**Comment by data-sync-user** [comment]
➤ Andrei Bodea commented:

Hello, please note that I was able to check this Experiment on v9000 here are my results so far: 

I had the following scenario:

  * Tab 1 open with www.wikipedia.com ( http://www.wikipedia.com )
  * Tab 2 open with www.google.com ( http://www.google.com )
  * Tab 3 open with www.youtube.com ( http://www.youtube.com )
After tapping on the home button (the last selected tab was 3rd one www.youtube.com ( http://www.youtube.com )) → tapping on the www.wikipedia.com ( http://www.wikipedia.com ) from shortcuts in order to open it, in the tabs tray please note that the redirect was done correctly and once I accessed the tabs tray I was able to see that Tab 1 was selected with Wikipedia link open.

In previous versions without this Experiment when selecting again wikipedia.com ( http://wikipedia.com ) from shortcuts that link would have been opened again in Tab 3 without any redirect.

Please note that I will do some more tests on this one and update here accordingly once I move forward and then I will also close the issue.

**Comment by data-sync-user** [comment]
➤ Andrei Bodea commented:

Hello, we have this new issue https://github.com/mozilla-mobile/firefox-ios/issues/20957 ( https://github.com/mozilla-mobile/firefox-ios/issues/20957|smart-link )  Norberto Andres Furlanis there any chance we can verify this with the design team and see if this is intended? 
Looks like an empty homepage tab is opened after every redirect and if using FF daily at the end of the day a user can get to way too many empty homepages.
